python/paddle/distributed/auto_parallel/intermediate/tensor_parallel.py

```diff
@@ -940,3 +943,402 @@ def tensor_parallel(model, optimizer=None, config=None):
+            cp_degree = process_mesh.shape[cp_index]
+            if isinstance(args, (list, tuple)):
+                all_args = []
+                for input_tensor in args:
+                    # check input_ids
+                    assert (
+                        input_tensor.is_dist()
+                    ), "Input tensor must be a distributed tensor."
+                    assert (
+                        len(input_tensor.shape) == 2
+                    ), f"input_ids should be [batch_size, seq_len], but got {input_tensor.shape}"
+                    placements = input_tensor.placements
+                    if placements is None:
+                        placements = [
+                            dist.Replicate()
+                            for _ in range(len(process_mesh.shape))
+                        ]
+                    assert (
+                        placements[cp_index] == dist.Replicate()
+                    ), "Input tensor must be a replicated tensor in cp mesh."
+                    reshard_input = shard_seq_load_balance(input_tensor, 1)
+                    all_args.append(reshard_input)
+                new_args = tuple(all_args)
+                return new_args
+            elif isinstance(args, paddle.Tensor):
+                reshard_input = shard_seq_load_balance(input_tensor, 1)
+                return reshard_input
+            else:
+                raise ValueError(
+                    f"Unsupported argument type: {type(args)}. Expected list of tensors or single tensor."
+                )
+
+        return p2p_split_input
+
+    def reshard_post_hook(self, process_mesh):
+        def reshard_hook(layer, input, output):
+            cp_index = process_mesh.dim_names.index('sep')
+            cp_degree = process_mesh.shape[cp_index]
+            placements = output.placements
+            assert (
+                output.is_dist()
+            ), f"output {output} must be a distributed tensor."
+            placements[cp_index] = dist.Replicate()
+            target_output = dist.reshard(output, process_mesh, placements)
+            return target_output
+
+        return reshard_hook
+
+    def apply(self, layer, process_mesh, shard_param_list):
+        if self.backend == 'all2all':
+            # Deepspeed Ulysses
+            layer.register_forward_pre_hook(
+                self.all2all_split_input_pre_hook(process_mesh)
+            )
+        elif self.backend == 'p2p':
+            # Ring FlashAttention
+            layer.register_forward_pre_hook(
+                self.p2p_split_input_pre_hook(process_mesh)
+            )
+        else:
+            logging.warning(
+                f'{self.backend} is not supported backend for context parallel'
+            )
+        if self.period == 'end':
+            layer.register_forward_post_hook(
+                self.reshard_post_hook(process_mesh)
+            )
+
+
+class ContextParallel(PlanBase):
```

Comment:
ContextParallel and TensorParallel are independent parallel strategies, it is recommended to put them in separate files later. e.g. context_parallel.py

Reply:
done